### PR TITLE
Use clippy.

### LINKED ISSUE #812: Lint rust code with clippy
I think we should use cargo-clippy. Does anybody oppose?

### PR DESCRIPTION
**Pull request by thomcc**
Fixes #812.

This is a large change-set but is almost entirely mechanical. Anything nontrivial I just added #[allow(clippy::blah)] with a FIXME. I'll file a followup for those but in bulk.

Note: `#[allow(unknown_lints)]` is only needed in crates that actually specify clippy lints, but I added it to everything just for completeness.

### CODE REVIEW

**Comment by thomcc** [comment]
I couldn't think of a name for the combined task so I kept them separate

File: components/places/src/storage/bookmarks/public_node.rs [hunk 1]
```diff
@@ -111,27 +111,43 @@ fn get_child_guids(db: &PlacesDb, parent: RowId) -> Result<Vec<SyncGuid>> {
...
 }
 
+enum ChildInfo {
+    NoChildren,
+    Guids(Vec<SyncGuid>),
+    Nodes(Vec<PublicNode>),
+}
+
+impl ChildInfo {
+    fn guids_nodes(self) -> (Option<Vec<SyncGuid>>, Option<Vec<PublicNode>>) {
```
Comment (linacambridge) [inline components/places/src/storage/bookmarks/public_node.rs]: I was going to suggest having `ChildInfo` hold the parent, and implementing `From<ChildInfo> for PublicNode` that does what `PublicNode::from(rb).with_children(child_guids, child_nodes)` does now...but that's more invasive, so your call.

File: components/places/src/storage/bookmarks/root_guid.rs [hunk 2]
```diff
@@ -61,15 +61,15 @@ impl BookmarkRootGuid {
         self.guid().clone()
     }
 
-    pub fn from_str(guid: &str) -> Option<Self> {
+    pub fn well_known(guid: &str) -> Option<Self> {
```
Comment (linacambridge) [inline components/places/src/storage/bookmarks/root_guid.rs]: Hmm, why didn't Clippy like `from_str`, but allowed `from_guid`? I liked the symmetry between `from_str` and `from_guid` before (and matching `as_str` and `as_guid`, too)...

File: components/places/src/types.rs [hunk 3]
```diff
@@ -15,16 +15,17 @@ use std::time::{Duration, SystemTime, UNIX_EPOCH};
 pub struct SyncGuid(pub String);
 
 impl SyncGuid {
+    #[allow(clippy::new_without_default)] // This probably should not be called `new`...
```
Comment (linacambridge) [inline components/places/src/types.rs]: Ehhh. I see where it's going with that, but implementing `Default` doesn't really make sense here, and `generate` is a mouthful. Sometimes Clippy is overeager.

File: components/push/communications/src/lib.rs [hunk 4]
```diff
@@ -246,7 +248,7 @@ impl Connection for ConnectHttp {
             )
             .send()
         {
-            Ok(_) => return Ok(true),
+            Ok(_) => Ok(true),
```
Comment (linacambridge) [inline components/push/communications/src/lib.rs]: Drive-by note: If we set up the `impl_from_error!` chain for push errors, I wonder if we could unify `reqwest::Error` with `CommunicationServerError`, and make this `.send().map(|_| true)`—but that's well outside the scope of this patch.

File: components/support/ffi/src/handle_map.rs [hunk 5]
```diff
@@ -685,6 +691,7 @@ unsafe impl IntoFfi for Handle {
 /// // pub extern "C" fn mylib_destroy_thing(h: u64, err: &mut ExternError)
 /// define_handle_map_deleter!(ITEMS, mylib_destroy_thing);
 /// ```
+#[derive(Default)]
 pub struct ConcurrentHandleMap<T> {
```
Comment (linacambridge) [inline components/support/ffi/src/handle_map.rs]: Can we implement `new` in terms of `ConcurrentHandleMap::default()`?

File: components/sync15/src/request.rs [hunk 6]
```diff
@@ -1323,14 +1324,28 @@ mod test {
         );
     }
 
+    macro_rules! assert_feq {
```
Comment (linacambridge) [inline components/sync15/src/request.rs]: Can we implement this comparison directly on `ServerTimestamp`? Looks like it came in handy in `bso_record.rs`, too...

File: components/sync15/src/telemetry.rs [hunk 7]
```diff
@@ -547,6 +547,12 @@ impl SyncTelemetry {
     }
 }
 
+impl Default for SyncTelemetry {
```
Comment (linacambridge) [inline components/sync15/src/telemetry.rs]: Can this be derived? I guess we'd have to implement `Default` for `Stopwatch`, too...

File: components/places/src/storage/bookmarks/public_node.rs (hunk 1 above)
Comment (thomcc) [reply components/places/src/storage/bookmarks/public_node.rs]: This was the minimum to make it stop complaining about the types being too complex, we already have issues on file to clean this up more.

File: components/places/src/storage/bookmarks/root_guid.rs (hunk 2 above)
Comment (thomcc) [reply components/places/src/storage/bookmarks/root_guid.rs]: It insisted either we implement `std::str::FromStr` or rename it if we're going to call things from_str.

AFAICT FromStr requires we return a Result<T, impl std::error::Error>, so I just renamed it.

File: components/places/src/types.rs (hunk 3 above)
Comment (thomcc) [reply components/places/src/types.rs]: I think renaming it SyncGuid::rand() would be reasonable

File: components/sync15/src/request.rs (hunk 6 above)
Comment (thomcc) [reply components/sync15/src/request.rs]: Honestly we want both of these to be direct comparisons, but eh. I actually don't like this clippy lint, and considered turning it off, but there's so little math in this repo that it's not worth it.

std::f64::EPSILON is actually wrong to use here if we wanted a relative comparison, so I don't think it should be shared. f64::EPSILON is the distance between 1.0 and the next float above 1.0. We expect the timestamps to be much larger than this, and floats get further apart as they get larger, so this is just not right.

If we actually wanted to compare for floats for approximate equality, we'd probably want something like `(t0 - t1).abs() <= THRESHOLD * t0.abs().max(t1.abs()).max(1.0)`, where THRESHOLD is something we would have to define (see http://realtimecollisiondetection.net/blog/?p=89 for some discussion on how you might derive this, but note that EPSILON there is not the same as std::f64::EPSILON).

However, in practice, we know ServerTimestamps are seconds with resolution to the hundredths place, and so *really* we should just convert to an integer to avoid such issues.

I'll file a followup for us to make ServerTimestamp store milliseconds instead of being a f64.

File: components/support/ffi/src/handle_map.rs (hunk 5 above)
Comment (thomcc) [reply components/support/ffi/src/handle_map.rs]: No, apparently the auto-derived Default has a T: Default bound. I've changed it to implement Default using new() though.

File: components/sync15/src/request.rs (hunk 6 above)
Comment (thomcc) [reply components/sync15/src/request.rs]: Filed #815
